### ISSUE TITLE
segfault while testing with pgbench (as of 700419d570bb34cd8e94239ffe1165cdc1fb0737)

### ISSUE THREAD

**Issue by tfoertsch** [pgbouncer/pgbouncer#97]
As of this commit:

commit 700419d570bb34cd8e94239ffe1165cdc1fb0737
Author: Marko Kreen <markokr@gmail.com>
Date:   Fri Nov 6 19:56:10 2015 +0200

    Postpone change of expect_rfq_count
    
    When changed too early and client socket is paused,
    then packet will processed again when socket is woken
    up and expect_rfq_count will be too high.

I managed to create a core dump. Here is a stack trace:

```
Program terminated with signal 11, Segmentation fault.
#0  0x0000000000416560 in sbuf_send_pending (sbuf=0x15e06b0) at src/sbuf.c:467
467             IOBuf *io = sbuf->io;
(gdb) bt
#0  0x0000000000416560 in sbuf_send_pending (sbuf=0x15e06b0) at src/sbuf.c:467
#1  0x00000000004167a6 in sbuf_process_pending (sbuf=0x15e06b0) at src/sbuf.c:545
#2  0x00000000004169d0 in sbuf_main_loop (skip_recv=<optimized out>, sbuf=0x15e06b0) at src/sbuf.c:709
#3  sbuf_main_loop (sbuf=0x15e06b0, skip_recv=<optimized out>) at src/sbuf.c:648
#4  0x00007f81743d8254 in event_base_loop () from /usr/lib/x86_64-linux-gnu/libevent-2.0.so.5
#5  0x000000000040f7e3 in main_loop_once () at src/main.c:699
#6  0x0000000000406804 in main (argc=<optimized out>, argv=<optimized out>) at src/main.c:858
(gdb) info threads
  Id   Target Id         Frame 
* 1    Thread 0x7f8174e78700 (LWP 2512) 0x0000000000416560 in sbuf_send_pending (sbuf=0x15e06b0) at src/sbuf.c:467
(gdb) print *sbuf
$3 = {ev = {ev_active_next = {tqe_next = 0x16a8f90, tqe_prev = 0x15bced0}, ev_next = {tqe_next = 0x0, tqe_prev = 0x15bf3f0}, ev_timeout_pos = {ev_next_with_common_timeout = {tqe_next = 0xffffffff, tqe_prev = 0x0}, min_heap_idx = -1}, ev_fd = 10, ev_base = 0x15bcc40, 
    _ev = {ev_io = {ev_io_next = {tqe_next = 0x15e06b0, tqe_prev = 0x15e06f0}, ev_timeout = {tv_sec = 0, tv_usec = 0}}, ev_signal = {ev_signal_next = {tqe_next = 0x15e06b0, tqe_prev = 0x15e06f0}, ev_ncalls = 0, ev_pncalls = 0x0}}, ev_events = 18, ev_res = 0, 
    ev_flags = 128, ev_pri = 0 '\000', ev_closure = 2 '\002', ev_timeout = {tv_sec = 209282, tv_usec = 970952}, ev_callback = 0x416ae0 <sbuf_recv_cb>, ev_arg = 0x15e06b0}, wait_type = 0 '\000', pkt_action = 2 '\002', sock = 0, pkt_remain = 1, 
  proto_cb = 0x4178f0 <server_proto>, dst = 0x0, io = 0x0, ops = 0x42f300, tls = 0x0, tls_host = 0x0}
(gdb) 
```

One line of code after that one generating the segfault there is an `AssertActive(sbuf)`. This would not have passed because `->sock==0`

pgbouncer was built on Debian with `--without-cares`. Connections to the client are established using unix domain sockets. Connections to the backend use SSL sockets. Pgbench ran for a little more than 2h before triggering the segfault.

```
/usr/lib/postgresql/9.3/bin/pgbench -U write -p 6432 -h /var/run/postgresql -c 60 -T 36000 pgbench
```

In parallel with the pgbench test I ran 30 jobs like this using 3 different databases:

```
while sleep $(dc <<<"3k $RANDOM 10000/p"); do
   echo '
       with a as (select pg_sleep(r.r) as dummy, r.r from random() r(r))
       select * from transaction.transaction cross join a;
   ';
done | psql -X -h /var/run/postgresql -p 6432 costarica-write write >/dev/null &
```

Transaction.transaction is a table containing 33 rows of test data,

Configuration:
[userlist.txt](https://github.com/pgbouncer/pgbouncer/files/49535/userlist.txt)
[pgbouncer.ini.txt](https://github.com/pgbouncer/pgbouncer/files/49539/pgbouncer.ini.txt)
[databases.ini-nopw.txt](https://github.com/pgbouncer/pgbouncer/files/49541/databases.ini-nopw.txt)

**Comment by markokr** [comment]
Did you compile with --enable-cassert?  If not the Assert-s are compiled out.

**Comment by markokr** [comment]
Could you print out the PgSocket struct that surrounds the sbuf?

I guess you need to take sizeof of both to calculate the PgSocket pointer.

**Comment by markokr** [comment]
I could reproduce it.  it's related to ssl connection establishment.  It's easier to reproduce if you make server_lifetime small.

sbuf_send_pending is being run on JUSTFREE socket.  Quick fix would be to make sbuf check for closed sockets, seems its not enough that only *_proto does it. But there is something else going on too, as disconnect is called in dubious circumstances.  I'll look further.
